src/main/java/com/github/firmwehr/gentle/firm/optimization/ConstantFolding.java

```diff
@@ -0,0 +1,588 @@
+package com.github.firmwehr.gentle.firm.optimization;
+
+import com.github.firmwehr.gentle.InternalCompilerException;
+import com.github.firmwehr.gentle.output.Logger;
+import firm.BackEdges;
+import firm.Graph;
+import firm.Mode;
+import firm.Program;
+import firm.Relation;
+import firm.TargetValue;
+import firm.bindings.binding_irgopt;
+import firm.bindings.binding_irnode;
+import firm.nodes.Add;
+import firm.nodes.And;
+import firm.nodes.Bad;
+import firm.nodes.Binop;
+import firm.nodes.Block;
+import firm.nodes.Cmp;
+import firm.nodes.Cond;
+import firm.nodes.Const;
+import firm.nodes.Conv;
+import firm.nodes.Div;
+import firm.nodes.Id;
+import firm.nodes.Member;
+import firm.nodes.Minus;
+import firm.nodes.Mod;
+import firm.nodes.Mul;
+import firm.nodes.Node;
+import firm.nodes.NodeVisitor;
+import firm.nodes.Not;
+import firm.nodes.Or;
+import firm.nodes.Phi;
+import firm.nodes.Proj;
+import firm.nodes.Return;
+import firm.nodes.Sel;
+import firm.nodes.Shl;
+import firm.nodes.Shr;
+import firm.nodes.Shrs;
+import firm.nodes.Size;
+import firm.nodes.Sub;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.function.BiConsumer;
+import java.util.function.BinaryOperator;
+import java.util.stream.StreamSupport;
+
+import static com.github.firmwehr.gentle.util.GraphDumper.dumpGraph;
+
+public class ConstantFolding extends NodeVisitor.Default {
+	private static final Logger LOGGER = new Logger(ConstantFolding.class);
+
+	private final Graph graph;
+	private final Map<Node, TargetValue> constants;
+	private final Deque<Node> worklist;
+	private final Map<Node, Node> replacements;
```

Comment:
```suggestion
	// stores potential replacements for each node, since changing them during iteration would break iterator
	private final Map<Node, Node> replacements;
```